File: packages/capp-sdk-ts/src/modules/corridors.ts
import { CAPPClient } from '../client';
import { CorridorStatus, CorridorEvent, CorridorFeedResponse } from '../models';
import { handleApiError } from '../utils';

export class CorridorsModule {
    constructor(private client: CAPPClient) { }

    public async status(corridor: string): Promise<CorridorStatus> {
        const res = await this.client.fetch(`/corridors/${corridor}/status`);
        await handleApiError(res);
        return res.json();
    }

    public async getFeed(corridor: string, lookbackDays: number = 7): Promise<CorridorFeedResponse> {
        const res = await this.client.fetch(`/corridors/feed?corridor=${encodeURIComponent(corridor)}&lookback_days=${lookbackDays}`);
        await handleApiError(res);
        return res.json();
    }

    public async list(): Promise<string[]> {
        const res = await this.client.fetch('/corridors');
        await handleApiError(res);
        const data = await res.json();
        return data.corridors || [];
    }

    public async *subscribe(corridor: string): AsyncGenerator<CorridorEvent, void, unknown> {
        const res = await this.client.fetch(`/corridors/${corridor}/events`);
        await handleApiError(res);

        if (!res.body) return;

        const reader = res.body.getReader();
        const decoder = new TextDecoder();
        let buffer = '';

        while (true) {
            const { value, done } = await reader.read();
            if (done) break;

            buffer += decoder.decode(value, { stream: true });
            const lines = buffer.split('\n');
            buffer = lines.pop() || '';

            for (const line of lines) {
                if (line.startsWith('data: ')) {
                    try {
                        yield JSON.parse(line.slice(6));
                    } catch (e) {
                        // Ignore parse errors on stream data
                    }
                }
            }
        }
    }
}

File: packages/capp-sdk/capp/modules/corridors.py
import json
import httpx
from typing import List, AsyncGenerator
from ..models import CorridorStatus, CorridorEvent, CorridorFeedResponse
from .._utils import handle_api_error

class CorridorsModule:
    def __init__(self, client: httpx.AsyncClient):
        self._client = client
        
    async def status(self, corridor: str) -> CorridorStatus:
        res = await self._client.get(f"/corridors/{corridor}/status")
        handle_api_error(res)
        return CorridorStatus(**res.json())

    async def get_feed(self, corridor: str, lookback_days: int = 7) -> CorridorFeedResponse:
        res = await self._client.get(f"/corridors/feed?corridor={corridor}&lookback_days={lookback_days}")
        handle_api_error(res)
        return CorridorFeedResponse(**res.json())

    async def list(self) -> List[str]:
        res = await self._client.get("/corridors")
        handle_api_error(res)
        return res.json().get("corridors", [])
        
    async def subscribe(self, corridor: str) -> AsyncGenerator[CorridorEvent, None]:
        async with self._client.stream("GET", f"/corridors/{corridor}/events") as response:
            handle_api_error(response)
            async for line in response.aiter_lines():
                if line.startswith("data: "):
                    data_str = line[6:]
                    try:
                        event_data = json.loads(data_str)
                        yield CorridorEvent(**event_data)
                    except json.JSONDecodeError:
                        continue

    async def subscribe_all(self) -> AsyncGenerator[CorridorEvent, None]:
        async with self._client.stream("GET", "/corridors/events") as response:
            handle_api_error(response)
            async for line in response.aiter_lines():
                if line.startswith("data: "):
                    data_str = line[6:]
                    try:
                        event_data = json.loads(data_str)
                        yield CorridorEvent(**event_data)
                    except json.JSONDecodeError:
                        continue
